update search/opt (#95)
OI-wiki/i18n: docs/search/opt.md

author: CBW2007, ChungZH, TrisolarisHD, abc1763613206, Ir1d

## Introduction

> DFS (Depth First Search) is a really common algorithm in  OI and most of the problems can be solved using it. However, in most cases, it can only get you some points, because the time complexity of DFS is extremely high and there are very few problems where a brute force DFS can be the most optimized solution. (If you have not studied DFS, please refer to [dfs](./dfs.md) for more details).

Reviewer: ```suggestion
DFS (Depth First Search) is a really common algorithm used in OI and most of the problems can be solved using it. However, in most cases, it can only get partial points, because the time complexity of DFS is extremely high and there are very few problems that can be solved with a brute-force DFS. (If you have not studied DFS, please refer to [dfs](./dfs.md) for more details).
```
- partial points

But how to get a few more points when DFS is not the best answer? This article will introduce some practical optimization algorithms (commonly known as "pruning").

> Let's first look at a section of DFS template, and the following templates will be modified on this basis.

Reviewer: modified based on it.

```cpp
int ans = worst_case, now;  // now is the current answer
void dfs(input_data) {
  if (destination_arrived) ans = the_best_from_the_current_and_existing_solutions;
  for (all_possible_solutions)
    if (feasible) {

>       carry_out_operation;

Reviewer: perform

      dfs(downsizing);
      withdraw_operation;
    }
}
```

Among them, ans can be the record of the solution, then the best one selected from the current and the existing solutions will become the output answer.

## Pruning methods

> There are three most commonly used pruning: memory search, optimal pruning, and feasible pruning.

Reviewer: memorized
